src/endianity.rs

```diff
@@ -11,103 +12,127 @@ use parser::{Error, Result};
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 2`.
+    #[inline]
+    fn read_u16(self, buf: &[u8]) -> u16 {
+        if self.is_big_endian() {
+            byteorder::BigEndian::read_u16(buf)
+        } else {
+            byteorder::LittleEndian::read_u16(buf)
+        }
     }
-}
 
-impl byteorder::ByteOrder for LittleEndian {
-    fn read_u16(buf: &[u8]) -> u16 {
-        byteorder::LittleEndian::read_u16(buf)
-    }
-    fn read_u32(buf: &[u8]) -> u32 {
-        byteorder::LittleEndian::read_u32(buf)
+    /// Reads an unsigned 32 bit integer from `buf`.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 4`.
+    fn read_u32(self, buf: &[u8]) -> u32 {
+        if self.is_big_endian() {
+            byteorder::BigEndian::read_u32(buf)
+        } else {
+            byteorder::LittleEndian::read_u32(buf)
+        }
     }
-    fn read_u64(buf: &[u8]) -> u64 {
-        byteorder::LittleEndian::read_u64(buf)
+
+    /// Reads an unsigned 64 bit integer from `buf`.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    fn read_u64(self, buf: &[u8]) -> u64 {
+        if self.is_big_endian() {
+            byteorder::BigEndian::read_u64(buf)
+        } else {
+            byteorder::LittleEndian::read_u64(buf)
+        }
     }
-    fn read_uint(buf: &[u8], nbytes: usize) -> u64 {
-        byteorder::LittleEndian::read_uint(buf, nbytes)
+
+    /// Reads a signed 16 bit integer from `buf`.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 2`.
+    #[inline]
+    fn read_i16(self, buf: &[u8]) -> i16 {
+        self.read_u16(buf) as i16
     }
-    fn write_u16(buf: &mut [u8], n: u16) {
-        byteorder::LittleEndian::write_u16(buf, n)
+
+    /// Reads a signed 32 bit integer from `buf`.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 4`.
+    #[inline]
+    fn read_i32(self, buf: &[u8]) -> i32 {
+        self.read_u32(buf) as i32
     }
-    fn write_u32(buf: &mut [u8], n: u32) {
-        byteorder::LittleEndian::write_u32(buf, n)
+
+    /// Reads a signed 64 bit integer from `buf`.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    #[inline]
+    fn read_i64(self, buf: &[u8]) -> i64 {
+        self.read_u64(buf) as i64
     }
-    fn write_u64(buf: &mut [u8], n: u64) {
-        byteorder::LittleEndian::write_u64(buf, n)
+
+    /// Writes an unsigned 64 bit integer `n` to `buf`.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    fn write_u64(self, buf: &mut [u8], n: u64) {
+        if self.is_big_endian() {
+            byteorder::BigEndian::write_u64(buf, n)
+        } else {
+            byteorder::LittleEndian::write_u64(buf, n)
+        }
     }
-    fn write_uint(buf: &mut [u8], n: u64, nbytes: usize) {
-        byteorder::LittleEndian::write_uint(buf, n, nbytes)
+}
+
+/// Little endian byte order.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct LittleEndian();
```

Comment:
This should do it:

```rust
#[cfg(target_endian = "little")]
pub use self::LittleEndian as NativeEndian;

#[cfg(target_endian = "big")]
pub use self::BigEndian as NativeEndian;
```

However, when I try this I start getting errors in `tests/parse_self.rs` (?!?!??!)

<details>

```
failures:

---- test_parse_self_debug_aranges stdout ----
	Reading section "debug_aranges" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_aranges"
thread 'test_parse_self_debug_aranges' panicked at 'Should parse arange OK: UnexpectedEof', src/libcore/result.rs:860
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: std::panicking::begin_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::expect
  10: parse_self::test_parse_self_debug_aranges
  11: <F as test::FnBox<T>>::call_box
  12: __rust_maybe_catch_panic

---- test_parse_self_debug_pubtypes stdout ----
	Reading section "debug_info" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_info"
Reading section "debug_abbrev" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_abbrev"
Reading section "debug_pubtypes" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_pubtypes"
thread 'test_parse_self_debug_pubtypes' panicked at 'Should parse pubtype OK: UnexpectedEof', src/libcore/result.rs:860
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: std::panicking::begin_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::expect
  10: parse_self::test_parse_self_debug_pubtypes
  11: <F as test::FnBox<T>>::call_box
  12: __rust_maybe_catch_panic

---- test_parse_self_debug_pubnames stdout ----
	Reading section "debug_info" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_info"
Reading section "debug_abbrev" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_abbrev"
Reading section "debug_pubnames" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_pubnames"
thread 'test_parse_self_debug_pubnames' panicked at 'Should parse pubname OK: UnexpectedEof', src/libcore/result.rs:860
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: std::panicking::begin_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::expect
  10: parse_self::test_parse_self_debug_pubnames
  11: <F as test::FnBox<T>>::call_box
  12: __rust_maybe_catch_panic

---- test_parse_self_eh_frame stdout ----
	Reading section "eh_frame" at path "/Users/fitzgen/src/gimli/fixtures/self/eh_frame"
thread 'test_parse_self_eh_frame' panicked at 'Should parse CFI entry OK: UnexpectedEof', src/libcore/result.rs:860
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: std::panicking::begin_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::expect
  10: parse_self::test_parse_self_eh_frame
  11: <F as test::FnBox<T>>::call_box
  12: __rust_maybe_catch_panic

---- test_parse_self_debug_line stdout ----
	Reading section "debug_info" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_info"
Reading section "debug_abbrev" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_abbrev"
Reading section "debug_line" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_line"
Reading section "debug_str" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_str"
thread 'test_parse_self_debug_line' panicked at 'Should parse compilation unit: UnexpectedEof', src/libcore/result.rs:860
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: std::panicking::begin_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::expect
  10: parse_self::test_parse_self_debug_line
  11: <F as test::FnBox<T>>::call_box
  12: __rust_maybe_catch_panic

---- test_parse_self_debug_info stdout ----
	Reading section "debug_info" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_info"
Reading section "debug_abbrev" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_abbrev"
thread 'test_parse_self_debug_info' panicked at 'Should parse compilation unit: UnexpectedEof', src/libcore/result.rs:860
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: std::panicking::begin_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::expect
  10: parse_self::test_parse_self_debug_info
  11: <F as test::FnBox<T>>::call_box
  12: __rust_maybe_catch_panic

---- test_parse_self_debug_ranges stdout ----
	Reading section "debug_info" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_info"
Reading section "debug_abbrev" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_abbrev"
Reading section "debug_ranges" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_ranges"
thread 'test_parse_self_debug_ranges' panicked at 'Should parse compilation unit: UnexpectedEof', src/libcore/result.rs:860
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: std::panicking::begin_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::expect
  10: parse_self::test_parse_self_debug_ranges
  11: <F as test::FnBox<T>>::call_box
  12: __rust_maybe_catch_panic

---- test_parse_self_debug_loc stdout ----
	Reading section "debug_info" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_info"
Reading section "debug_abbrev" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_abbrev"
Reading section "debug_loc" at path "/Users/fitzgen/src/gimli/fixtures/self/debug_loc"
thread 'test_parse_self_debug_loc' panicked at 'Should parse compilation unit: UnexpectedEof', src/libcore/result.rs:860
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: std::panicking::begin_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::expect
  10: parse_self::test_parse_self_debug_loc
  11: <F as test::FnBox<T>>::call_box
  12: __rust_maybe_catch_panic


failures:
    test_parse_self_debug_aranges
    test_parse_self_debug_info
    test_parse_self_debug_line
    test_parse_self_debug_loc
    test_parse_self_debug_pubnames
    test_parse_self_debug_pubtypes
    test_parse_self_debug_ranges
    test_parse_self_eh_frame
```

</details>